# MC316 Final Exam – Part B: PYTHON

* You will have 1 hour and 40 minutes to finish the test.
* This part includes 10 questions. Each question is worth 15 points.
* You get extra 30 points for full answers.
* Please read the instructions carefully. 

## I. SET UP


### Import pandas library for use in scripting. 

In [1]:
import pandas as pd

### Import the datasets and name them.
For this test, you will analyze revenue and expenditure data released by Kansas Department of Administration for the fiscal year 2018. While the original data can be directly downloaded [here](http://kanview.ks.gov/DataDownload.aspx), for your convenience, the instructor had them uploaded to a server at the following urls. 

#### For Kansas State University:
1. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv
2. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv

#### For University of Kansas:
3. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv
4. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv

You will need to import all 4 datasets and name them in an easy way to call out later.

*Please refer to the [Glossary](http://kanview.ks.gov/Glossary.aspx#f) for data description.

In [2]:
ExpKSU2018 = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv")

In [3]:
RevKSU2018 = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv")

In [42]:
ExpKU2018 = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv")

In [41]:
RevKU2018 = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv")

## II. REQUIRED QUESTIONS
* Make sure you type the answers in the cells right below the codes.
* If you need to add extra cells, click the + button in the toolbar up the top. 

### Q1: What is the total revenue recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total revenue recorded for University of Kansas in the same year?

In [10]:
RevKSU2018.Total.sum()

854324643.84

In [44]:
RevKU2018.Total.sum()

808767902.55

<b>Answer: The total revenue recorded for KSU in 2018 was 854324643.84.  The total revenue recorded for KU in 2018 was 808767902.55 </b> 

### Q2: What is the total expense recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total expense recorded for University of Kansas in the same year?

In [11]:
ExpKSU2018.Total.sum()

1018584373.02

In [43]:
ExpKU2018.Total.sum()

990238875.6099999

<b>Answer: The total expense recorded for KSU in 2018 was 1018584373.02.  The total expense recorded for KU in 2018 was 990238875.6099999</b> 

### Q3: Where did the two universities' primary revenues come from? 
*Refer to the description [here](http://kanview.ks.gov/Revenues/AcctCodeRevenues.aspx) to write your answer.

In [15]:
RevKSU2018.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   637
REV FROM USE OF MONEY & PROP      354
REVENUE TRANSFERS                 341
NON REVENUE RECEIPTS              259
OTHER REVENUE                     160
GIFT DONATION FEDERAL GRANTS      152
TAXES                               7
Name: Primary_Account_Code_Descr, dtype: int64

In [19]:
RevKU2018.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   376
REV FROM USE OF MONEY & PROP      251
NON REVENUE RECEIPTS              157
REVENUE TRANSFERS                 153
OTHER REVENUE                      57
TAXES                              45
GIFT DONATION FEDERAL GRANTS       38
Name: Primary_Account_Code_Descr, dtype: int64

<b>Answer: Agency earnings was the Primary revenue for both universities. </b> 

### Q4: What was the biggest source of revenues for Kansas State University? Compare to the one of University of Kansas.

Remove hashtag and run the following code to suppress scientific notation:

In [20]:
pd.set_option('display.float_format', lambda x: '%.3f' % x) 

In [58]:
RevKSU2018.groupby("Primary_Account_Code_Descr").Total.sum().reset_index()

Primary_Account_Code_Descr         Total
0               AGENCY EARNINGS 412565112.820
1  GIFT DONATION FEDERAL GRANTS 278778848.030
2          NON REVENUE RECEIPTS  71947564.050
3                 OTHER REVENUE   8432030.160
4  REV FROM USE OF MONEY & PROP  34139318.040
5             REVENUE TRANSFERS  48458616.490
6                         TAXES      3154.250

In [57]:
RevKU2018.groupby("Primary_Account_Code_Descr").Total.sum().reset_index()

Primary_Account_Code_Descr         Total
0  AGENCY EARNINGS                408409337.320
1  GIFT DONATION FEDERAL GRANTS   193836668.320
2  NON REVENUE RECEIPTS            69753247.670
3  OTHER REVENUE                   68241930.500
4  REV FROM USE OF MONEY & PROP    40950779.540
5  REVENUE TRANSFERS               26013855.400
6  TAXES                            1562083.800

In [33]:
412565112.820 - 408409337.320

4155775.5

<b>Answer:Agency Earnings was the highest source revenue for both universites and the difference between the two is 4155775.5</b> 

### Q5: What accounts contributed to Kansas State University's biggest source of revenues?
If you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [35]:
RevKSU2018["Primary_Account_Code_Descr"] = RevKSU2018["Primary_Account_Code_Descr"].str.strip()

After cleaning the spaces, you can start filtering your datasets to find the answer for Q5.

In [38]:
RevKSU2018.groupby("Primary_Account_Code_Descr").Total.sum().reset_index().sort_values("Total", ascending=False)

Primary_Account_Code_Descr         Total
0               AGENCY EARNINGS 412565112.820
1  GIFT DONATION FEDERAL GRANTS 278778848.030
2          NON REVENUE RECEIPTS  71947564.050
5             REVENUE TRANSFERS  48458616.490
4  REV FROM USE OF MONEY & PROP  34139318.040
3                 OTHER REVENUE   8432030.160
6                         TAXES      3154.250

<b>Answer: Agency Earnings Accounts Contributed the most.</b>

### Q6: Where did the money go? (Answer for both universities)
*Refer to the description [here] (http://kanview.ks.gov/Expenditures/AcctCodeExpenditures.aspx) to write your answer.

In [72]:
ExpKSU2018.groupby("Primary_Account_Code_Descr").Total.sum().reset_index().sort_values("Total", ascending=False)

Primary_Account_Code_Descr         Total
7  SALARIES AND WAGES             476367053.010
6  NONEXPENSE ITEMS               130529992.790
2  CONTRACTUAL SERVICES           128541598.280
5  GRANTS CLAIMS SHARED REVENUE   106551248.640
0  CAPITAL OUTLAY                  64350550.130
1  COMMODITIES                     42670908.840
4  EXPENSE TRANSFERS               37130576.230
3  DEBT SERVICE                    32442445.100

<b>Answer:observe graphs about to see the allocation of money.</b> 

In [66]:
ExpKU2018.groupby("Primary_Account_Code_Descr").Total.sum().reset_index().sort_values("Total", ascending=False)

Primary_Account_Code_Descr         Total
7  SALARIES AND WAGES             471996016.150
6  NONEXPENSE ITEMS               161145084.400
2  CONTRACTUAL SERVICES           139630507.780
5  GRANTS CLAIMS SHARED REVENUE    71439705.360
0  CAPITAL OUTLAY                  70676545.250
4  EXPENSE TRANSFERS               25815417.520
3  DEBT SERVICE                    25309742.030
1  COMMODITIES                     24225857.120

### Q7: What costs two universities the most? 

In [55]:
ExpKSU2018.groupby("Primary_Account_Code_Descr").Total.sum().reset_index().sort_values("Total", ascending=False)

Primary_Account_Code_Descr         Total
7  SALARIES AND WAGES             476367053.010
6  NONEXPENSE ITEMS               130529992.790
2  CONTRACTUAL SERVICES           128541598.280
5  GRANTS CLAIMS SHARED REVENUE   106551248.640
0  CAPITAL OUTLAY                  64350550.130
1  COMMODITIES                     42670908.840
4  EXPENSE TRANSFERS               37130576.230
3  DEBT SERVICE                    32442445.100

In [56]:
ExpKU2018.groupby("Primary_Account_Code_Descr").Total.sum().reset_index().sort_values("Total", ascending=False)

Primary_Account_Code_Descr         Total
7  SALARIES AND WAGES             471996016.150
6  NONEXPENSE ITEMS               161145084.400
2  CONTRACTUAL SERVICES           139630507.780
5  GRANTS CLAIMS SHARED REVENUE    71439705.360
0  CAPITAL OUTLAY                  70676545.250
4  EXPENSE TRANSFERS               25815417.520
3  DEBT SERVICE                    25309742.030
1  COMMODITIES                     24225857.120

<b>Answer: Salaries and Wages cost the two Universities the most.</b> 

### Q8: Calculate the percentage of this cost over each university's total expense.

In [59]:
ExpKSU2018.Total.sum()

1018584373.02

In [62]:
(476367053.010 / 1018584373.02) * 100

46.76755952946929

In [63]:
ExpKU2018.Total.sum()

990238875.6099999

In [64]:
(471996016.150 / 990238875.6099999) * 100

47.66486428431163

<b>Answer: KSU's Salary and Wages made up around 46.8% of the total expenditure. KU's Salary and Wages made up around 47.7% of the total expenditure. </b> 

### Q9: Breakdown the cost you found in Q7 for both universities.
Again, if you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [3]:
#yourdataframe["yourcolumnname"] = yourdataframe["yourcolumnname"].str.strip()

<b>Answer:</b>

### Q10: Compare the breakdown costs from Q9 side by side.
When you merge the two subsets, make sure to add the attribute <b>how ="outer"</b> so that you can keep every item from both sets.
*For example: merge = pd.merge(df1, df2, how="outer", on="columnname")

<b>Answer:</b>

## III. BONUS QUESTIONS
* Each question's worth 15 points.


### B1: Which university made a profit in the fiscal year of 2018?

In [50]:
RevKSU2018.Total.sum() - ExpKSU2018.Total.sum()

-164259729.17999995

In [46]:
RevKU2018.Total.sum() - ExpKU2018.Total.sum()

-181470973.05999994

<b>Answer: Neither University made a profit in 2018</b>

### B2: Which software program do you like the most in this course? Why?

<b>Answer:I enjoyed open tableau the most. It was easiest to figure out and i made some oretty cool graphics with it.  </b>

### END OF THE TEST